### PR TITLE
On mismatched argument count point at arguments

### PR DESCRIPTION
**Pull request by estebank**

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -3751,8 +3751,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    *span,
+                    if args.is_empty() {
+                        // `sp` doesn't point at the whole `foo.bar()`, only at `bar`.
+                        vec![]
+                    } else {
+                        // Skip the receiver.
+                        args[1..].iter().map(|arg| arg.span).collect::<Vec<Span>>()
+                    },
+                ),
+                // The following should never happen?
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]: Insert a `span_bug!` then?

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -3762,6 +3794,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 ),
                 DiagnosticId::Error(error_code.to_owned()),
             );
+            let label = format!("supplied {}", potentially_plural_count(arg_count, "parameter"));
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
            let label = format!("supplied {}", potentially_plural_count(arg_count, "argument"));
```
And in comments as well.

File: src/librustc_typeck/check/mod.rs [hunk 3]
```diff
@@ -3751,8 +3751,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        // ^^^ -  -  - supplied 3 parameters
+                        // |
+                        // expected 2 parameters
+                        args.iter().map(|arg| arg.span).collect::<Vec<Span>>()
+                    },
+                ),
+                hir::ExprKind::MethodCall(_, span, args) => (
+                    *span,
+                    if args.is_empty() {
+                        // `sp` doesn't point at the whole `foo.bar()`, only at `bar`.
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]: Weird. Perhaps it *should* point to the whole `foo.bar()`?
Then we'll be able to use the same `vec![tcx.sess.source_map().next_point(*span).with_hi(sp.hi())]` as `ExprKind::Call`.

File: src/librustc_typeck/check/mod.rs [hunk 4]
```diff
@@ -3761,17 +3761,64 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
                                  c_variadic: bool,
                                  sugg_unit: bool| {
+            let (span, arg_spans) = match &expr.kind {
+                hir::ExprKind::Call(hir::Expr { span, .. }, args) => (
+                    *span,
+                    if args.is_empty() {
+                        // foo()
+                        // ^^^-- supplied 0 arguments
+                        // |
+                        // expected 2 parameters
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
                        // expected 2 arguments
```

File: src/librustc_typeck/check/mod.rs [hunk 5]
```diff
@@ -3761,17 +3761,64 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        vec![tcx.sess.source_map().next_point(*span).with_hi(sp.hi())]
+                    } else {
+                        // foo(1, 2, 3)
+                        // ^^^ -  -  - supplied 3 arguments
+                        // |
+                        // expected 2 parameters
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
                        // expected 2 arguments
```

File: src/librustc_typeck/check/mod.rs [hunk 6]
```diff
@@ -3761,17 +3761,64 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                ),
+                k => span_bug!(sp, "checking argument types on a non-call: `{:?}`", k),
+            };
             let mut err = tcx.sess.struct_span_err_with_code(
-                sp,
+                span,
                 &format!(
                     "this function takes {}{} but {} {} supplied",
                     if c_variadic { "at least " } else { "" },
                     potentially_plural_count(expected_count, "parameter"),
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
                    potentially_plural_count(expected_count, "argument"),
```

File: src/librustc_typeck/check/mod.rs [hunk 7]
```diff
@@ -3761,17 +3761,64 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                  error_code: &str,
                                  c_variadic: bool,
                                  sugg_unit: bool| {
+            let (span, arg_spans) = match &expr.kind {
+                hir::ExprKind::Call(hir::Expr { span, .. }, args) => (
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]: Besides the number of elements to skip and one span, the logic is common between `ExprKind::Call` and `ExprKind::MethodCall` and can be moved to outside of the match.

Comment (estebank) [reply src/librustc_typeck/check/mod.rs]: done

**Comment by petrochenkov** [comment]
r=me after squashing the commits

**Comment by JohnTitor** [comment]
Seems failed in https://github.com/rust-lang/rust/pull/69085#issuecomment-585061151
@bors r-

**Comment by estebank** [comment]
Blocked on https://github.com/rust-lang/cargo/pull/7883

**Comment by estebank** [comment]
After https://github.com/rust-lang/rust/pull/69265, @bors r=petrochenkov

**Pull request merged**